addons/l10n_my_edi/models/account_move.py

```diff
@@ -0,0 +1,661 @@
+        self._l10n_my_edi_check_can_update_status()
+
+        successfully_updated_invoices = self.env['account.move']
+        for document in self:
+            result = proxy_user._l10n_my_edi_contact_proxy(
+                endpoint='api/l10n_my_edi/1/update_status',
+                params={
+                    'status_values': {
+                        'uuid': document.l10n_my_edi_external_uuid,
+                        'reason': reason,
+                        'status': status,
+                    },
+                },
+            )
+
+            # If it is not a success, it will have raised an error.
+            if 'error' in result:
+                self._message_log(body=self._l10n_my_edi_map_error(result['error']))
+            else:
+                successfully_updated_invoices |= document
+
+        if status in self._fields['l10n_my_edi_state'].get_values(self.env):
+            successfully_updated_invoices._l10n_my_edi_set_status(
+                state=status,
+                message=_('This invoice has been %(status)s for reason: %(reason)s', status=status, reason=reason),
+            )
+
+        if self._can_commit():
+            self._cr.commit()
+
+    @api.model
+    def _cron_l10n_my_edi_synchronize_myinvois(self):
+        """
+        This cron is based on the recommended method to fetch the status of the documents according to their doc.
+        record_count can be used to define how many submission to process per cron.
+        """
+        # We will leave a 3s interval between each api call to make sure we don't get throttled.
+        # One api call will be done for each submission (a submission may contain multiple invoices if sent in batch).
+
+        # First step is to get the invoices for which the status is not yet final.
+        # A invoice whose status will not change anymore is: (cancelled or invalid) or has been validated more than 74h ago.
+        # /!\ when an invoice validation is pending, l10n_my_edi_validation_time is still None. These also need to be updated.
+        datetime_threshold = datetime.datetime.now() - datetime.timedelta(hours=74)
+        invoices = self.env['account.move'].search([
+            ('l10n_my_edi_state', 'not in', (False, 'invalid', 'cancelled')),  # In practice we could ignore rejected invoices as we don't fully support vendor bills.
+            '|', ('l10n_my_edi_validation_time', '>', datetime_threshold),
+            ('l10n_my_edi_validation_time', '=', None),
+        ])
+        # Check if we have any invoices to process, otherwise we can skip everything else.
+        if not invoices:
+            return
+
+        invoices_per_company = invoices.grouped('company_id')
+        # Use _notify_progress to ensure that we continue if all batches have not been done in time..
+        total_submissions_to_process = len(invoices.mapped('l10n_my_edi_submission_uid'))
+        submission_processed = 0
+        for company in set(invoices_per_company.keys()):
```

Comment:
- `invoices_per_company.keys()` is already a set-like object
- You could directly iterate on `for company, company_invoices in invoices_per_company.items()`